File: algorithms/216.Combination_Sum_III/Combination_Sum_III.java
public class Solution {
  public List<List<Integer>> combinationSum3(int k, int n) {
    List<List<Integer>> ans = new ArrayList<>();
    if (k == 0) {
      return ans;
    }

    dfsHelper(k, n, 1, 0, new ArrayList<>(), ans);
    return ans;
  }

  private void dfsHelper(int k, int n, int pos, int sum, List<Integer> taken, List<List<Integer>> ans) {
    if (taken.size() == k || sum > n) {
      if (sum == n) {
        ans.add(new ArrayList<>(taken));
      }
      return ;
    }

    for (int i = pos; i <= 9; i++) {
      sum += i;
      taken.add(i);
      dfsHelper(k, n, i + 1, sum, taken, ans);
      taken.remove(taken.size() - 1);
      sum -= i;
    }
  }
}

File: algorithms/216.Combination_Sum_III/Combination_Sum_III.py
class Solution:
    # @param {integer} k
    # @param {integer} n
    # @return {integer[][]}
    def combinationSum3(self, k, n):
        result = []
        self.combinationSumRecu(result, [], 1, k, n)
        return result
    
    def combinationSumRecu(self, result, intermediate, start, k, target):
        if k == 0 and target == 0:
            result.append(list(intermediate))
        elif k < 0:
            return
        while start < 10 and start * k + k * (k - 1) / 2 <= target:
            intermediate.append(start)
            self.combinationSumRecu(result, intermediate, start + 1, k - 1, target - start)
            intermediate.pop()
            start += 1
class Solution:
    # @param {integer} k
    # @param {integer} n
    # @return {integer[][]}
    def combinationSum3(self, k, n):
        if n > sum([i for i in range(1, 11)]):
            return []

        res = []
        self.sum_help(k, n, 1, [], res)
        return res


    def sum_help(self, k, n, curr, arr, res):
        if len(arr) == k:
            if sum(arr) == n:
                res.append(list(arr))
            return

        if len(arr) > k or curr > 9:
            return

        for i in range(curr, 10):
            arr.append(i)
            self.sum_help(k, n, i + 1, arr, res)
            arr.pop()
